README.md
# Html5-QRCode-ru

## Кроссплатформенный сканер QR-кодов и штрих-кодов HTML5

Используйте эту легковесную библиотеку, чтобы легко и быстро интегрировать QR-код, штрих-код и другие распространенные возможности сканирования кодов в ваше веб-приложение.

## Изменения
- Переведено на русский язык
- Обновлены зависимости

## Credits
Основано на базе https://www.npmjs.com/package/html5-qrcode


changelog.md
### Version 2.3.8-6
- Обновлён zxing-js до v0.20.0

### Version 2.3.8-5
- Переведено на русский язык
- Обновлены зависимости
- Убраны лишние строки (не, они действительно лишние)

### Version 2.3.8
 -   `Scanner Paused` color tone changes.
 -   Misc code refactor.
 -   Backwards compatibility improvement by [bilal-elchami@](https://github.com/bilal-elchami) in [PR#700](https://github.com/mebjas/html5-qrcode/pull/700).
 -   Documentation generated and moved to [scanapp.org/html5-qrcode-docs](https://scanapp.org/html5-qrcode-docs/)
 -   `Html5QrcodeScannerConfig` exported, thanks to [PR#728](https://github.com/mebjas/html5-qrcode/pull/728) by [seyfer@](https://github.com/seyfer).
 -   `LitExample` added, thanks to [PR#704](https://github.com/mebjas/html5-qrcode/pull/704) by [kusigit@](https://github.com/kusigit).

output/tests/src/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clip = exports.isNullOrUndefined = exports.BaseLoggger = exports.Html5QrcodeErrorFactory = exports.Html5QrcodeErrorTypes = exports.Html5QrcodeResultFactory = exports.QrcodeResultFormat = exports.Html5QrcodeConstants = exports.Html5QrcodeScanType = exports.isValidHtml5QrcodeSupportedFormats = exports.DecodedTextType = exports.Html5QrcodeSupportedFormats = void 0;
var Html5QrcodeSupportedFormats;
(function (Html5QrcodeSupportedFormats) {
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["QR_CODE"] = 0] = "QR_CODE";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["AZTEC"] = 1] = "AZTEC";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["CODABAR"] = 2] = "CODABAR";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["CODE_39"] = 3] = "CODE_39";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["CODE_93"] = 4] = "CODE_93";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["CODE_128"] = 5] = "CODE_128";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["DATA_MATRIX"] = 6] = "DATA_MATRIX";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["MAXICODE"] = 7] = "MAXICODE";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["ITF"] = 8] = "ITF";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["EAN_13"] = 9] = "EAN_13";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["EAN_8"] = 10] = "EAN_8";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["PDF_417"] = 11] = "PDF_417";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["RSS_14"] = 12] = "RSS_14";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["RSS_EXPANDED"] = 13] = "RSS_EXPANDED";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["UPC_A"] = 14] = "UPC_A";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["UPC_E"] = 15] = "UPC_E";
    Html5QrcodeSupportedFormats[Html5QrcodeSupportedFormats["UPC_EAN_EXTENSION"] = 16] = "UPC_EAN_EXTENSION";
})(Html5QrcodeSupportedFormats || (exports.Html5QrcodeSupportedFormats = Html5QrcodeSupportedFormats = {}));
const html5QrcodeSupportedFormatsTextMap = new Map([
    [Html5QrcodeSupportedFormats.QR_CODE, "QR_CODE"],
    [Html5QrcodeSupportedFormats.AZTEC, "AZTEC"],
    [Html5QrcodeSupportedFormats.CODABAR, "CODABAR"],
    [Html5QrcodeSupportedFormats.CODE_39, "CODE_39"],
    [Html5QrcodeSupportedFormats.CODE_93, "CODE_93"],
    [Html5QrcodeSupportedFormats.CODE_128, "CODE_128"],
    [Html5QrcodeSupportedFormats.DATA_MATRIX, "DATA_MATRIX"],
    [Html5QrcodeSupportedFormats.MAXICODE, "MAXICODE"],
    [Html5QrcodeSupportedFormats.ITF, "ITF"],
    [Html5QrcodeSupportedFormats.EAN_13, "EAN_13"],
    [Html5QrcodeSupportedFormats.EAN_8, "EAN_8"],
    [Html5QrcodeSupportedFormats.PDF_417, "PDF_417"],
    [Html5QrcodeSupportedFormats.RSS_14, "RSS_14"],
    [Html5QrcodeSupportedFormats.RSS_EXPANDED, "RSS_EXPANDED"],
    [Html5QrcodeSupportedFormats.UPC_A, "UPC_A"],
    [Html5QrcodeSupportedFormats.UPC_E, "UPC_E"],
    [Html5QrcodeSupportedFormats.UPC_EAN_EXTENSION, "UPC_EAN_EXTENSION"]
]);
var DecodedTextType;
(function (DecodedTextType) {
    DecodedTextType[DecodedTextType["UNKNOWN"] = 0] = "UNKNOWN";
    DecodedTextType[DecodedTextType["URL"] = 1] = "URL";
})(DecodedTextType || (exports.DecodedTextType = DecodedTextType = {}));
function isValidHtml5QrcodeSupportedFormats(format) {
    return Object.values(Html5QrcodeSupportedFormats).includes(format);
}
exports.isValidHtml5QrcodeSupportedFormats = isValidHtml5QrcodeSupportedFormats;
var Html5QrcodeScanType;
(function (Html5QrcodeScanType) {
    Html5QrcodeScanType[Html5QrcodeScanType["SCAN_TYPE_CAMERA"] = 0] = "SCAN_TYPE_CAMERA";
    Html5QrcodeScanType[Html5QrcodeScanType["SCAN_TYPE_FILE"] = 1] = "SCAN_TYPE_FILE";
})(Html5QrcodeScanType || (exports.Html5QrcodeScanType = Html5QrcodeScanType = {}));
class Html5QrcodeConstants {
}
exports.Html5QrcodeConstants = Html5QrcodeConstants;
Html5QrcodeConstants.GITHUB_PROJECT_URL = "https://github.com/mebjas/html5-qrcode";
Html5QrcodeConstants.SCAN_DEFAULT_FPS = 2;
Html5QrcodeConstants.DEFAULT_DISABLE_FLIP = false;
Html5QrcodeConstants.DEFAULT_REMEMBER_LAST_CAMERA_USED = true;
Html5QrcodeConstants.DEFAULT_SUPPORTED_SCAN_TYPE = [
    Html5QrcodeScanType.SCAN_TYPE_CAMERA,
    Html5QrcodeScanType.SCAN_TYPE_FILE
];
class QrcodeResultFormat {
    constructor(format, formatName) {
        this.format = format;
        this.formatName = formatName;
    }
    toString() {
        return this.formatName;
    }
    static create(format) {
        if (!html5QrcodeSupportedFormatsTextMap.has(format)) {
            throw `${format} not in html5QrcodeSupportedFormatsTextMap`;
        }
        return new QrcodeResultFormat(format, html5QrcodeSupportedFormatsTextMap.get(format));
    }
}
exports.QrcodeResultFormat = QrcodeResultFormat;
class Html5QrcodeResultFactory {
    static createFromText(decodedText) {
        let qrcodeResult = {
            text: decodedText
        };
        return {
            decodedText: decodedText,
            result: qrcodeResult
        };
    }
    static createFromQrcodeResult(qrcodeResult) {
        return {
            decodedText: qrcodeResult.text,
            result: qrcodeResult
        };
    }
}
exports.Html5QrcodeResultFactory = Html5QrcodeResultFactory;
var Html5QrcodeErrorTypes;
(function (Html5QrcodeErrorTypes) {
    Html5QrcodeErrorTypes[Html5QrcodeErrorTypes["UNKWOWN_ERROR"] = 0] = "UNKWOWN_ERROR";
    Html5QrcodeErrorTypes[Html5QrcodeErrorTypes["IMPLEMENTATION_ERROR"] = 1] = "IMPLEMENTATION_ERROR";
    Html5QrcodeErrorTypes[Html5QrcodeErrorTypes["NO_CODE_FOUND_ERROR"] = 2] = "NO_CODE_FOUND_ERROR";
})(Html5QrcodeErrorTypes || (exports.Html5QrcodeErrorTypes = Html5QrcodeErrorTypes = {}));
class Html5QrcodeErrorFactory {
    static createFrom(error) {
        return {
            errorMessage: error,
            type: Html5QrcodeErrorTypes.UNKWOWN_ERROR
        };
    }
}
exports.Html5QrcodeErrorFactory = Html5QrcodeErrorFactory;
class BaseLoggger {
    constructor(verbose) {
        this.verbose = verbose;
    }
    log(message) {
        if (this.verbose) {
            console.log(message);
        }
    }
    warn(message) {
        if (this.verbose) {
            console.warn(message);
        }
    }
    logError(message, isExperimental) {
        if (this.verbose || isExperimental === true) {
            console.error(message);
        }
    }
    logErrors(errors) {
        if (errors.length === 0) {
            throw "Logger#logError called without arguments";
        }
        if (this.verbose) {
            console.error(errors);
        }
    }
}
exports.BaseLoggger = BaseLoggger;
function isNullOrUndefined(obj) {
    return (typeof obj === "undefined") || obj === null;
}
exports.isNullOrUndefined = isNullOrUndefined;
function clip(value, minValue, maxValue) {
    if (value > maxValue) {
        return maxValue;
    }
    if (value < minValue) {
        return minValue;
    }
    return value;
}
exports.clip = clip;
//# sourceMappingURL=core.js.map

output/tests/src/ui/scanner/file-selection-ui.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSelectionUi = void 0;
const strings_1 = require("../../strings");
const base_1 = require("./base");
class FileSelectionUi {
    constructor(parentElement, showOnRender, onFileSelected) {
        this.fileBasedScanRegion = this.createFileBasedScanRegion();
        this.fileBasedScanRegion.style.display
            = showOnRender ? "block" : "none";
        parentElement.appendChild(this.fileBasedScanRegion);
        let fileScanLabel = document.createElement("label");
        fileScanLabel.setAttribute("for", this.getFileScanInputId());
        fileScanLabel.style.display = "inline-block";
        this.fileBasedScanRegion.appendChild(fileScanLabel);
        this.fileSelectionButton
            = base_1.BaseUiElementFactory.createElement("button", base_1.PublicUiElementIdAndClasses.FILE_SELECTION_BUTTON_ID);
        this.setInitialValueToButton();
        this.fileSelectionButton.addEventListener("click", (_) => {
            fileScanLabel.click();
        });
        fileScanLabel.append(this.fileSelectionButton);
        this.fileScanInput
            = base_1.BaseUiElementFactory.createElement("input", this.getFileScanInputId());
        this.fileScanInput.type = "file";
        this.fileScanInput.accept = "image/*";
        this.fileScanInput.style.display = "none";
        fileScanLabel.appendChild(this.fileScanInput);
        let $this = this;
        this.fileScanInput.addEventListener("change", (e) => {
            if (e == null || e.target == null) {
                return;
            }
            let target = e.target;
            if (target.files && target.files.length === 0) {
                return;
            }
            let fileList = target.files;
            const file = fileList[0];
            let fileName = file.name;
            $this.setImageNameToButton(fileName);
            onFileSelected(file);
        });
        let dragAndDropMessage = this.createDragAndDropMessage();
        this.fileBasedScanRegion.appendChild(dragAndDropMessage);
        this.fileBasedScanRegion.addEventListener("dragenter", function (event) {
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionActiveBorder();
            event.stopPropagation();
            event.preventDefault();
        });
        this.fileBasedScanRegion.addEventListener("dragleave", function (event) {
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionDefaultBorder();
            event.stopPropagation();
            event.preventDefault();
        });
        this.fileBasedScanRegion.addEventListener("dragover", function (event) {
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionActiveBorder();
            event.stopPropagation();
            event.preventDefault();
        });
        this.fileBasedScanRegion.addEventListener("drop", function (event) {
            event.stopPropagation();
            event.preventDefault();
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionDefaultBorder();
            var dataTransfer = event.dataTransfer;
            if (dataTransfer) {
                let files = dataTransfer.files;
                if (!files || files.length === 0) {
                    return;
                }
                let isAnyFileImage = false;
                for (let i = 0; i < files.length; ++i) {
                    let file = files.item(i);
                    if (!file) {
                        continue;
                    }
                    let imageType = /image.*/;
                    if (!file.type.match(imageType)) {
                        continue;
                    }
                    isAnyFileImage = true;
                    let fileName = file.name;
                    $this.setImageNameToButton(fileName);
                    onFileSelected(file);
                    dragAndDropMessage.innerText
                        = strings_1.Html5QrcodeScannerStrings.dragAndDropMessage();
                    break;
                }
                if (!isAnyFileImage) {
                    dragAndDropMessage.innerText
                        = strings_1.Html5QrcodeScannerStrings
                            .dragAndDropMessageOnlyImages();
                }
            }
        });
    }
    hide() {
        this.fileBasedScanRegion.style.display = "none";
        this.fileScanInput.disabled = true;
    }
    show() {
        this.fileBasedScanRegion.style.display = "block";
        this.fileScanInput.disabled = false;
    }
    isShowing() {
        return this.fileBasedScanRegion.style.display === "block";
    }
    resetValue() {
        this.fileScanInput.value = "";
        this.setInitialValueToButton();
    }
    createFileBasedScanRegion() {
        let fileBasedScanRegion = document.createElement("div");
        fileBasedScanRegion.style.textAlign = "center";
        fileBasedScanRegion.style.margin = "auto";
        fileBasedScanRegion.style.width = "80%";
        fileBasedScanRegion.style.maxWidth = "600px";
        fileBasedScanRegion.style.border
            = this.fileBasedScanRegionDefaultBorder();
        fileBasedScanRegion.style.padding = "10px";
        fileBasedScanRegion.style.marginBottom = "10px";
        return fileBasedScanRegion;
    }
    fileBasedScanRegionDefaultBorder() {
        return "6px dashed #ebebeb";
    }
    fileBasedScanRegionActiveBorder() {
        return "6px dashed rgb(153 151 151)";
    }
    createDragAndDropMessage() {
        let dragAndDropMessage = document.createElement("div");
        dragAndDropMessage.innerText
            = strings_1.Html5QrcodeScannerStrings.dragAndDropMessage();
        dragAndDropMessage.style.fontWeight = "400";
        return dragAndDropMessage;
    }
    setImageNameToButton(imageFileName) {
        const MAX_CHARS = 20;
        this.fileSelectionButton.innerText = strings_1.Html5QrcodeScannerStrings.fileSelectionChooseAnother();
    }
    setInitialValueToButton() {
        this.fileSelectionButton.innerText = strings_1.Html5QrcodeScannerStrings.fileSelectionChooseImage();
    }
    getFileScanInputId() {
        return "html5-qrcode-private-filescan-input";
    }
    static create(parentElement, showOnRender, onFileSelected) {
        return new FileSelectionUi(parentElement, showOnRender, onFileSelected);
    }
}
exports.FileSelectionUi = FileSelectionUi;
//# sourceMappingURL=file-selection-ui.js.map

src/strings.ts
/**
 * @fileoverview
 * Strings used by {@class Html5Qrcode} & {@class Html5QrcodeScanner}
 * 
 * The word "QR Code" is registered trademark of DENSO WAVE INCORPORATED
 * http://www.denso-wave.com/qrcode/faqpatent-e.html
 */

/**
 * Strings used in {@class Html5Qrcode}.
 */
export class Html5QrcodeStrings {

    public static codeParseError(exception: any): string {
        return `Ошибка разбора QR-кода, ошибка = ${exception}`;
    }

    public static errorGettingUserMedia(error: any): string {
        return `Ошибка получения пользовательского медиа, ошибка = ${error}`;
    }

    public static onlyDeviceSupportedError(): string {
        return "Устройство не поддерживает navigator.mediaDevices, поддерживаемый в этом случае параметр cameraIdOrConfig - это параметр deviceId (строка).";
    }

    public static cameraStreamingNotSupported(): string {
        return "Воспроизведение видео с камеры не поддерживается браузером.";
    }

    public static unableToQuerySupportedDevices(): string {
        return "Невозможно получить доступ к поддерживаемым устройствам, неизвестная ошибка.";
    }

    public static insecureContextCameraQueryError(): string {
        return "Доступ к камере поддерживается только в защищенном контексте, таком как https или localhost.";
    }

    public static scannerPaused(): string {
        return "Сканер приостановлен";
    }
}

/**
 * Strings used in {@class Html5QrcodeScanner}.
 */
export class Html5QrcodeScannerStrings {

    public static scanningStatus(): string {
        return "Сканирование";
    }

    public static idleStatus(): string {
        return "Ожидание";
    }

    public static errorStatus(): string {
        return "Ошибка";
    }

    public static permissionStatus(): string {
        return "Разрешение";
    }

    public static noCameraFoundErrorStatus(): string {
        return "Камеры не найдены";
    }

    public static lastMatch(decodedText: string): string {
        return `Последнее совпадение: ${decodedText}`;
    }

    public static codeScannerTitle(): string {
        return "Сканер QR-кода";
    }

    public static cameraPermissionTitle(): string {
        return "Запрос разрешения на использование камеры";
    }

    public static cameraPermissionRequesting(): string {
        return "Запрос разрешения на использование камеры...";
    }

    public static noCameraFound(): string {
        return "Камера не найдена";
    }

    public static scanButtonStopScanningText(): string {
        return "Остановить сканирование";
    }

    public static scanButtonStartScanningText(): string {
        return "Начать сканирование";
    }

    public static torchOnButton(): string {
        return "Включить фонарик";
    }

    public static torchOffButton(): string {
        return "Выключить фонарик";
    }

    public static torchOnFailedMessage(): string {
        return "Не удалось включить фонарик";
    }

    public static torchOffFailedMessage(): string {
        return "Не удалось выключить фонарик";
    }

    public static scanButtonScanningStarting(): string {
        return "Запуск камеры...";
    }

    /**
     * Text to show when camera scan is selected.
     *
     * This will be used to switch to file based scanning.
     */
    public static textIfCameraScanSelected(): string {
        return "Сканировать файл с изображением";
    }

    /**
     * Text to show when file based scan is selected.
     *
     * This will be used to switch to camera based scanning.
     */
    public static textIfFileScanSelected(): string {
        return "Сканировать с помощью камеры";
    }

    public static selectCamera(): string {
        return "Выберите камеру";
    }

    public static fileSelectionChooseImage(): string {
        return "Выберите изображение";
    }

    public static fileSelectionChooseAnother(): string {
        return "Выберите другое изображение";
    }

    public static fileSelectionNoImageSelected(): string {
        return "Изображение не выбрано";
    }

    /** Prefix to be given to anonymous cameras. */
    public static anonymousCameraPrefix(): string {
        return "Анонимная камера";
    }

    public static dragAndDropMessage(): string {
        return "Или перетащите изображение для сканирования";
    }

    public static dragAndDropMessageOnlyImages(): string {
        return "Или перетащите изображение для сканирования (не поддерживаются другие файлы)";
    }

    /** Value for zoom. */
    public static zoom(): string {
        return "увеличение";
    }

    public static loadingImage(): string {
        return "Загрузка изображения...";
    }

    public static cameraScanAltText(): string {
        return "Сканирование с помощью камеры";
    }

    public static fileScanAltText(): string {
        return "Сканирование с помощью файла";
    }
}

/** Strings used in {@class LibraryInfoDiv} */
export class LibraryInfoStrings {

    public static poweredBy(): string {
        return "Работает на основе ";
    }

    public static reportIssues(): string {
        return "Сообщить об ошибке";
    }
}


src/ui/scanner/file-selection-ui.ts
/**
 * @fileoverview
 * File for file selection UI handling.
 * 
 * @author mebjas <minhazav@gmail.com>
 * 
 * The word "QR Code" is registered trademark of DENSO WAVE INCORPORATED
 * http://www.denso-wave.com/qrcode/faqpatent-e.html
 */

import {Html5QrcodeScannerStrings} from "../../strings";
import {BaseUiElementFactory, PublicUiElementIdAndClasses} from "./base";

/**
 * Interface for callback when a file is selected by user using the button.
 */
export type OnFileSelected = (file: File) => void;

/** UI class for file selection handling. */
export class FileSelectionUi {

    private readonly fileBasedScanRegion: HTMLDivElement;
    private readonly fileScanInput: HTMLInputElement;
    private readonly fileSelectionButton: HTMLButtonElement;

    /** Creates object and renders. */
    private constructor(
        parentElement: HTMLDivElement,
        showOnRender: boolean,
        onFileSelected: OnFileSelected) {
        this.fileBasedScanRegion = this.createFileBasedScanRegion();
        this.fileBasedScanRegion.style.display
            = showOnRender ? "block" : "none";
        parentElement.appendChild(this.fileBasedScanRegion);

        let fileScanLabel = document.createElement("label");
        fileScanLabel.setAttribute("for", this.getFileScanInputId());
        fileScanLabel.style.display = "inline-block";

        this.fileBasedScanRegion.appendChild(fileScanLabel);
        
        this.fileSelectionButton
            = BaseUiElementFactory.createElement<HTMLButtonElement>(
                "button",
                PublicUiElementIdAndClasses.FILE_SELECTION_BUTTON_ID);
        this.setInitialValueToButton();

        // Bind click events with the label element.
        this.fileSelectionButton.addEventListener("click", (_) => {
            fileScanLabel.click();
        });
        fileScanLabel.append(this.fileSelectionButton);

        this.fileScanInput
            = BaseUiElementFactory.createElement<HTMLInputElement>(
                "input", this.getFileScanInputId());
        this.fileScanInput.type = "file";
        this.fileScanInput.accept = "image/*";
        this.fileScanInput.style.display = "none";
        fileScanLabel.appendChild(this.fileScanInput);
        
        let $this = this;
        /*eslint complexity: ["error", 5]*/
        this.fileScanInput.addEventListener("change", (e: Event) => {
            if (e == null || e.target == null) {
                return;
            }
            let target: HTMLInputElement = e.target as HTMLInputElement;
            if (target.files && target.files.length === 0) {
                return;
            }
            let fileList: FileList = target.files!;
            const file: File = fileList[0];
            let fileName = file.name;
            $this.setImageNameToButton(fileName);

            onFileSelected(file);
        });

        // Render drag and drop label
        let dragAndDropMessage = this.createDragAndDropMessage();
        this.fileBasedScanRegion.appendChild(dragAndDropMessage);

        this.fileBasedScanRegion.addEventListener("dragenter", function(event) {
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionActiveBorder();

            event.stopPropagation();
            event.preventDefault();
        });

        this.fileBasedScanRegion.addEventListener("dragleave", function(event) {
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionDefaultBorder();

            event.stopPropagation();
            event.preventDefault();
        });

        this.fileBasedScanRegion.addEventListener("dragover", function(event) {
            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionActiveBorder();

            event.stopPropagation();
            event.preventDefault();
        });

        /*eslint complexity: ["error", 10]*/
        this.fileBasedScanRegion.addEventListener("drop", function(event) {
            event.stopPropagation();
            event.preventDefault();

            $this.fileBasedScanRegion.style.border
                = $this.fileBasedScanRegionDefaultBorder();

            var dataTransfer = event.dataTransfer;
            if (dataTransfer) {
                let files = dataTransfer.files;
                if (!files || files.length === 0) {
                    return;
                }
                let isAnyFileImage = false;
                for (let i = 0; i < files.length; ++i) {
                    let file = files.item(i);
                    if (!file) {
                        continue;
                    }
                    let imageType = /image.*/;

                    // Only process images.
                    if (!file.type.match(imageType)) {
                        continue;
                    }

                    isAnyFileImage = true;
                    let fileName = file.name;
                    $this.setImageNameToButton(fileName);

                    onFileSelected(file);
                    dragAndDropMessage.innerText
                        = Html5QrcodeScannerStrings.dragAndDropMessage();
                    break;
                }
                
                // None of the files were images.
                if (!isAnyFileImage) {
                    dragAndDropMessage.innerText
                        = Html5QrcodeScannerStrings
                            .dragAndDropMessageOnlyImages();
                }
            }

        });
    }

    //#region Public APIs.
    /** Hide the file selection UI. */
    public hide() {
        this.fileBasedScanRegion.style.display = "none";
        this.fileScanInput.disabled = true;
    }

    /** Show the file selection UI. */
    public show() {
        this.fileBasedScanRegion.style.display = "block";
        this.fileScanInput.disabled = false;
    }

    /** Returns {@code true} if UI container is displayed. */
    public isShowing(): boolean {
        return this.fileBasedScanRegion.style.display === "block";
    }

    /** Reset the file selection value */
    public resetValue() {
        this.fileScanInput.value = "";
        this.setInitialValueToButton();
    }
    //#endregion

    //#region private APIs
    private createFileBasedScanRegion(): HTMLDivElement {
        let fileBasedScanRegion = document.createElement("div");
        fileBasedScanRegion.style.textAlign = "center";
        fileBasedScanRegion.style.margin = "auto";
        fileBasedScanRegion.style.width = "80%";
        fileBasedScanRegion.style.maxWidth = "600px";
        fileBasedScanRegion.style.border
            = this.fileBasedScanRegionDefaultBorder();
        fileBasedScanRegion.style.padding = "10px";
        fileBasedScanRegion.style.marginBottom = "10px";
        return fileBasedScanRegion;
    }

    private fileBasedScanRegionDefaultBorder() {
        return "6px dashed #ebebeb";
    }

    /** Border when a file is being dragged over the file scan region. */
    private fileBasedScanRegionActiveBorder() {
        return "6px dashed rgb(153 151 151)";
    }

    private createDragAndDropMessage(): HTMLDivElement {
        let dragAndDropMessage = document.createElement("div");
        dragAndDropMessage.innerText
            = Html5QrcodeScannerStrings.dragAndDropMessage();
        dragAndDropMessage.style.fontWeight = "400";
        return dragAndDropMessage;
    }

    private setImageNameToButton(imageFileName: string) {
        const MAX_CHARS = 20;
        if (imageFileName.length > MAX_CHARS) {
            // Strip first 8
            // Strip last 8
            // Add 4 dots
            let start8Chars = imageFileName.substring(0, 8);
            let length = imageFileName.length;
            let last8Chars = imageFileName.substring(length - 8, length);
        }

        this.fileSelectionButton.innerText = Html5QrcodeScannerStrings.fileSelectionChooseAnother();
    }

    private setInitialValueToButton() {
        this.fileSelectionButton.innerText = Html5QrcodeScannerStrings.fileSelectionChooseImage();
    }

    private getFileScanInputId(): string {
        return "html5-qrcode-private-filescan-input";
    }
    //#endregion

    /**
     * Creates a file selection UI and renders.
     * 
     * @param parentElement parent div element to render the UI to.
     * @param showOnRender if {@code true}, the UI will be shown upon render
     *  else hidden.
     * @param onFileSelected callback to be called when file selection changes.
     * 
     * @returns Instance of {@code FileSelectionUi}.
     */
    public static create(
        parentElement: HTMLDivElement,
        showOnRender: boolean,
        onFileSelected: OnFileSelected): FileSelectionUi {
        return new FileSelectionUi(
            parentElement, showOnRender, onFileSelected);
    }
} 
